ARROW-14761: [Doc] Helping with documentation (#11823)
apache/arrow: docs/source/developers/guide/documentation.rst

**************************
Helping with documentation
**************************

A great way to contribute to the project is to improve
documentation. 

If you are an Arrow user and you found some docs to be
incomplete or inaccurate, share your hard-earned knowledge
with the rest of the community.

If you haven't come across an error in the documentation
it is hard to find one and going through the docs in search

> for corrections is not a good way to go. In this case you can

Reviewer: This first sentence is a bit hard to follow. Maybe adding a comma after the first line would already help. 

Maybe just saying "If you didn't come across something to improve in the documentation itself, you can search for an issue ..." is sufficient?

Author: Yes, you are quite right. It's a confusing sentence. Will correct, thanks!

search for an issue that is dealing with the documentation

> on JIRA.

Reviewer 2: Perhaps add something about the "good-first-issue" and "good-second-issue" labels? Example search:
https://issues.apache.org/jira/issues/?jql=project%20%3D%20ARROW%20AND%20resolution%20%3D%20Unresolved%20AND%20labels%20in%20(good-first-issue%2C%20good-second-issue)%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC

Reviewer 2: Oh, actually it's done in https://github.com/apache/arrow/pull/11807 . Nevermind.

.. note::

> 	When searching for JIRA issue that deals with documentation

Reviewer 2: It seems your editor is configured to use tab characters for indentation, can you change it to only insert space characters?

(also, conventionally, the indentation size for `.rst` files is 3 spaces :-))

Author: Yeah, I got bitten by this yesterday! It should be correctly set in the new commit.

   try selecting **Components** from **More** tab in JIRA search
   and there find **Documentation** from the list.

>    See `Example search. <https://issues.apache.org/jira/issues/jql=project%20%3D%20ARROW%20AND%20status%20%3D%20Open%20AND%20resolution%20%3D%20Unresolved%20AND%20component%20%3D%20Documentation%20AND%20assignee%20in%20(EMPTY)%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC>`_

Reviewer: ```suggestion
   See `Example search. <https://issues.apache.org/jira/issues/?jql=project%20%3D%20ARROW%20AND%20status%20%3D%20Open%20AND%20resolution%20%3D%20Unresolved%20AND%20component%20%3D%20Documentation%20AND%20assignee%20in%20(EMPTY)%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC>`_
```

(otherwise the link doesn't work for me)

> 	See `Example search. <https://issues.apache.org/jira/browse/ARROW-14280?jql=project%20%3D%20ARROW%20AND%20status%20%3D%20Open%20AND%20resolution%20%3D%20Unresolved%20AND%20component%20%3D%20Documentation%20AND%20assignee%20in%20(EMPTY)%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC>`_

Reviewer 2: This URL points to a specific issue, we should give the link to the search summary instead, e.g. https://issues.apache.org/jira/issues/?jql=project%20%3D%20ARROW%20AND%20status%20%3D%20Open%20AND%20resolution%20%3D%20Unresolved%20AND%20component%20%3D%20Documentation%20AND%20assignee%20in%20(EMPTY)%20ORDER%20BY%20priority%20DESC%2C%20updated%20DESC

> .. figure:: jira_doc_1.jpeg

Reviewer 2: Should these figures perhaps be indented so as to be displayed inside the note?

   :scale: 40 %
   :alt: selecting Components in JIRA search

   First select Components tab in JIRA.

.. figure:: jira_doc_2.jpeg
   :scale: 40 %
   :alt: selecting Documentation in JIRA search

   Then choose Documentation from the Components list.

Documentation improvements are also a great way to gain some
experience with our submission and review process without
requiring a lot of local development environment setup. 

.. note::
	Many documentation-only changes can be made directly in the
	GitHub web interface by clicking the **edit** icon. This
	will handle making a fork and a pull request for you.

> .. figure:: github_edit_1.jpeg

Reviewer 2: Same suggestion here (put those inside the note).

**************************
Helping with documentation
**************************

**A great way to contribute to the project is to improve
documentation.**

If you are an Arrow user and you found some docs to be
incomplete or inaccurate, share your hard-earned knowledge
with the rest of the community.

If you didn't come across something to improve in the
documentation itself, you can search for an issue on JIRA.

.. note::

>    When searching for JIRA issue that deals with documentation
>    try selecting **Components** from **More** tab in JIRA search
>    and there find **Documentation** from the list.

Reviewer: ```suggestion
   When searching for JIRA issue that deals with documentation,
   try selecting **Components** from the **More** tab in JIRA search
   and select **Documentation** from the list.
```

      :scale: 40 %
      :alt: selecting Components in JIRA search

      First select Components tab in JIRA.

   .. figure:: jira_search_documentation.jpeg
      :scale: 40 %
      :alt: selecting Documentation in JIRA search

      Then choose Documentation from the Components list.

Documentation improvements are also a great way to gain some
experience with our submission and review process without
requiring a lot of local development environment setup. 

.. note::
   Many documentation-only changes can be made directly in the
   GitHub web interface by clicking the **Edit this page**
   on the right top corner of the documentations page. This
   will handle making a fork and a pull request for you.

   .. figure:: edit_page.jpeg
      :scale: 20 %
      :alt: click on edit this page

      On the right corner of the file in GitHub click on pen icon.

   .. figure:: github_edit_page.jpeg
      :scale: 30 %
      :alt: edit file in GitHub.

      Now you can edit the file in GitHub.

You could also build the entire project, make the change locally on
your branch and make the PR this way. But it is by no means superior
to simply editing via GitHub.

> If you wish to build the documentation also, follow detailed instructions

Reviewer: ```suggestion
If you wish to build the documentation locally, follow detailed instructions
```

on :ref:`building-docs`.

Where to find the correct file to change?
-----------------------------------------

Most of the documentation is located in the ``docs/source`` of the Arrow

> library. Source folder includes:

Reviewer: Now you added this list, I think it is relevant to mention "docstrings" for Python as well -> for editing the reference docs, you will need to edit the docstring the pyarrow package itself

Reviewer: And actually for C++ it's the same issue

Author: Ok, will do.

- **C++ documentation** section: ``docs/source/cpp``.
- **Development** section: ``docs/source/developers``.
- **Specificatons and protocols** section: ``docs/source/format``.
- Other language documentation: **C (GLib), Java, JavaScript** and **Python**
  in folders named by the language, for example: ``docs/source/c_glib``.

> - The documentation for the **R language** is located in the ``r/sub-directory``.

Reviewer: ```suggestion
- The documentation for the **R language** is located in the ``r/`` sub-directory.
```

?

   on the right top corner of the documentations page. This
   will handle making a fork and a pull request for you.

   .. figure:: edit_page.jpeg
      :scale: 20 %
      :alt: click on edit this page

      On the right corner of the file in GitHub click on pen icon.

   .. figure:: github_edit_page.jpeg
      :scale: 30 %
      :alt: edit file in GitHub.

      Now you can edit the file in GitHub.

You could also build the entire project, make the change locally on
your branch and make the PR this way. But it is by no means superior
to simply editing via GitHub.

If you wish to build the documentation locally, follow detailed instructions
on :ref:`building-docs`.

Where to find the correct file to change?
-----------------------------------------

Most of the documentation is located in the ``docs/source`` of the Arrow
library. Source folder includes:

- **C++ documentation** section: ``docs/source/cpp``.
- **Development** section: ``docs/source/developers``.
- **Specificatons and protocols** section: ``docs/source/format``.
- Other language documentation: **C (GLib), Java, JavaScript** and **Python**
  in folders named by the language, for example: ``docs/source/c_glib``.

> - The documentation for the **R language** is located in the ``r/`` sub-directory.

Author: @thisisnic did I get this R part of the documentation source right?

  - In the ``R/vignettes`` you can find the **Articles**, for example
    `Working with Cloud Storage (S3) <https://arrow.apache.org/docs/r/articles/fs.html>`_.
  - In the ``r/R`` you can edit reference documentation for example 
    in the ``r/R/dataset.R`` you can find the ``docstrings`` for the
    `open_dataset <https://arrow.apache.org/docs/r/reference/open_dataset.html>`_.

- **Reference documentation**

  To edit `API reference documentation for Python <https://arrow.apache.org/docs/python/api.html>`_,

>   you will need to edit the ``docstring`` of the ``pyarrow`` package itself.

Reviewer: ```suggestion
  you will need to edit the docstrings of the functions and methods located in the ``pyarrow`` package itself.
```
